apis/autoscaling/v1alpha1/mongodb_webhook.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright AppsCode Inc. and Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"errors"
+
+	dbapi "kubedb.dev/apimachinery/apis/kubedb/v1alpha2"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var mongoLog = logf.Log.WithName("mongodb-autoscaler")
+
+func (in *MongoDBAutoscaler) SetupWebhookWithManager(mgr manager.Manager) error {
+	return builder.WebhookManagedBy(mgr).
+		For(in).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/mutate-autoscaling-kubedb-com-v1alpha1-mongodbautoscaler,mutating=true,failurePolicy=fail,sideEffects=None,groups=autoscaling.kubedb.com,resources=mongodbautoscaler,verbs=create;update,versions=v1alpha1,name=mmongodbautoscaler.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &MongoDBAutoscaler{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (in *MongoDBAutoscaler) Default() {
+	mongoLog.Info("defaulting", "name", in.Name)
+}
+
+func (in *MongoDBAutoscaler) SetDefaults(db *dbapi.MongoDB) {
+	if in.Spec.Storage != nil {
+		setDefaultStorageValues(in.Spec.Storage.Standalone)
+		setDefaultStorageValues(in.Spec.Storage.ReplicaSet)
+		setDefaultStorageValues(in.Spec.Storage.Shard)
+		setDefaultStorageValues(in.Spec.Storage.ConfigServer)
+	}
+
+	if in.Spec.Compute != nil {
+		setDefaultComputeValues(in.Spec.Compute.Standalone)
+		setInMemoryDefaults(in.Spec.Compute.Standalone, db.Spec.StorageEngine)
+		setDefaultComputeValues(in.Spec.Compute.ReplicaSet)
+		setInMemoryDefaults(in.Spec.Compute.ReplicaSet, db.Spec.StorageEngine)
+		setDefaultComputeValues(in.Spec.Compute.Shard)
+		setInMemoryDefaults(in.Spec.Compute.Shard, db.Spec.StorageEngine)
+		setDefaultComputeValues(in.Spec.Compute.ConfigServer)
+		setInMemoryDefaults(in.Spec.Compute.ConfigServer, db.Spec.StorageEngine)
+		setDefaultComputeValues(in.Spec.Compute.Mongos)
+		setInMemoryDefaults(in.Spec.Compute.Mongos, db.Spec.StorageEngine)
+	}
+}
```

Comment:
Maybe keep 2 separate function for the ones need db and the one that doesn't need db. Then call the without db one from Default. Same for validator too.

Reply:
splitting up the validateFields func is not possible as db is being used in the `if-else` condition.